### PR TITLE
Use criterion for benchmarking

### PR DESCRIPTION
**Pull request by koushiro**
# What have I changed?

1.  Use `criterion` for benchmarking
2. Add `rustc-hex` and `faster-hex` for benchmarking
3. Add format and clippy check

### CODE REVIEW

File: benches/hex.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use rustc_hex::{FromHex, ToHex};
+
+const DATA: &[u8] = include_bytes!("../src/lib.rs");
+
+fn bench_encode(c: &mut Criterion) {
+    c.bench_function("hex_encode", |b| b.iter(|| black_box(hex::encode(DATA))));
```
Comment (LukasKalbertodt) [inline benches/hex.rs]: Note that criterion puts everything that is returned by the closure into a `black_box` already: https://docs.rs/criterion/0.2.11/src/criterion/lib.rs.html#335-345

So many of these `black_box`es can be removed. Of course, one might still prefer to keep them for explicitness.

Comment (koushiro) [reply benches/hex.rs]: > Note that criterion puts everything that is returned by the closure into a `black_box` already: https://docs.rs/criterion/0.2.11/src/criterion/lib.rs.html#335-345
> 
> So many of these `black_box`es can be removed. Of course, one might still prefer to keep them for explicitness.

Thanks for your reviewing :)
I will remove that.

**Comment by koushiro** [comment]
@KokaKiwi PTAL :)

File: benches/hex.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    c.bench_function("rustc_hex_decode", |b| {
+        let hex = DATA.to_hex::<String>();
+        b.iter(|| black_box(hex.from_hex::<Vec<u8>>().unwrap()))
+    });
+
+    c.bench_function("faster_hex_decode", move |b| {
+        let hex = faster_hex::hex_string(DATA).unwrap();
+        let len = DATA.len();
+        b.iter(|| {
+            let mut dst = Vec::with_capacity(len);
```
[the remark this answers is not in the document]
Comment (koushiro) [reply benches/hex.rs]: In each iteration, I think that the memory allocation and hex encoding should be considered as a complete operation to obtain the encoding result.

[the remark this answers is not in the document]
Comment (KokaKiwi) [reply benches/hex.rs]: Personally I don't think memory allocation should count in the encoding bench result, as it's something kind of out of scope of `hex`
By that I mean memory allocation benchs should stay where memory allocation are implemented, aka not here.

[the remark this answers is not in the document]
Comment (koushiro) [reply benches/hex.rs]: In my opinion, if you don't allocate the corresponding memory, you can't get the encoding result. Just like in the `hex` crate, you have to allocate a String as a result of hex encoding.

[the remark this answers is not in the document]
Comment (KokaKiwi) [reply benches/hex.rs]: Ah yeah forgot about the `encode` returning `String`, seems ok then, sorry ^^

**Comment by koushiro** [comment]
BTW, I recommend `squash and merge`, which will make the commit history clearer :)

**Pull request merged**